openwisp_firmware_upgrader/base/models.py

```diff
@@ -266,6 +296,32 @@ def create_upgrade_operation(self, batch):
         transaction.on_commit(lambda: upgrade_firmware.delay(operation.pk))
         return operation
 
+    @classmethod
+    def auto_add_device_firmware_to_device(cls, instance, created, **kwargs):
+        # Automatically associate DeviceFirmware to the registered Device
+        if created:
+            if not instance.device.os or not instance.device.model:
+                return
+            if instance.device.model not in REVERSE_FIRMWARE_IMAGE_MAP:
+                return
+
+            DeviceFirmware = load_model('DeviceFirmware')
+            FirmwareImage = load_model('FirmwareImage')
+
+            try:
+                image = FirmwareImage.objects.get(
+                    build__category__organization_id=instance.device.organization_id,
+                    build__os_identifier=instance.device.os,
+                    type=REVERSE_FIRMWARE_IMAGE_MAP[instance.device.model],
+                )
+                device_fw = DeviceFirmware(
```

Comment:
I think we can create a class method for this operation:

`DeviceFirmware.create_for_device(device)`, this method can execute this code.

I would also check if the `DeviceFirmware` already exists. It should not happen, but we never know.. and I get these kind of weird corner cases from time to time so now I want to start to avoid them to present themselves in the first place instead of having to debug them later on which is harder because we'll be busy doing other things.

So this method, before trying to create a new `DeviceFirmware`, we could do:

```python
qs = DeviceFirmware.objects.filter(device=dev)
if qs.exists():
    return qs.get()
```

We can then call this method in the celery task and in the migration.